src/core/net/tcp6.cpp

```diff
@@ -166,15 +293,113 @@ Error Tcp::Listener::StopListening(void)
+    SetNext(nullptr);
+
+exit:
+    return error;
 }
 
-Error Tcp::ProcessReceivedSegment(Message &aMessage, MessageInfo &aMessageInfo)
+Error Tcp::HandleMessage(ot::Ip6::Header &aIp6Header, Message &aMessage, MessageInfo &aMessageInfo)
 {
+    OT_UNUSED_VARIABLE(aIp6Header);
     OT_UNUSED_VARIABLE(aMessage);
     OT_UNUSED_VARIABLE(aMessageInfo);
 
-    return kErrorNotImplemented;
+    Error error = kErrorNotImplemented;
+
+    for (Endpoint *active = mEndpoints.GetHead(); active != nullptr; active = active->GetNext())
+    {
+    }
+
+    for (Listener *passive = mListeners.GetHead(); passive != nullptr; passive = passive->GetNext())
+    {
+    }
+
+    return error;
+}
+
+void Tcp::HandleTimer(Timer &aTimer)
+{
+    OT_ASSERT(&aTimer == &aTimer.GetInstance().Get<Tcp>().mTimer);
+    otLogDebgTcp("Main TCP timer expired");
+    aTimer.GetInstance().Get<Tcp>().FireAllPendingTimers();
+}
+
+void Tcp::FireAllPendingTimers()
+{
+    TimeMilli now                        = TimerMilli::GetNow();
+    bool      pendingTimer               = false;
+    TimeMilli earliestPendingTimerExpiry = now.GetDistantFuture();
+    Endpoint *endpoint                   = mEndpoints.GetHead();
+
+    OT_ASSERT(!mTimer.IsRunning());
+
+    /*
+     * The timer callbacks could potentially set/reset/cancel timers.
+     * Importantly, Endpoint::SetTimer and Endpoint::CancelTimer do not call
+     * this function to recompute the timer. If they did, we'd have a
+     * re-entrancy problem, where the callbacks called in this function could
+     * wind up re-entering this function in a nested call frame.
+     *
+     * In general, calling this function from Endpoint::SetTimer and
+     * Endpoint::CancelTimer could be inefficient, since those functions are
+     * called multiple times on each received TCP segment. If we want to
+     * prevent the main timer from firing except when an actual TCP timer
+     * expires, a better alternative is to reset the main timer in
+     * HandleMessage, right before processing signals. That would achieve that
+     * objective while avoiding re-entrancy issues altogether.
+     */
+
+    while (endpoint != nullptr)
+    {
+        TimeMilli expiry;
+        bool      hasFutureTimer;
+
+        if (endpoint->FirePendingTimers(now, hasFutureTimer, expiry))
+        {
+            /*
+             * If a non-OpenThread callback is called --- which, in practice,
+             * happens if the connection times out and the user-defined
+             * connection lost callback is called --- then we might have to
+             * start over. The reason is that the user might deinitialize
+             * endpoints, changing the structure of the linked list. For
+             * example, if the user deinitializes both this endpoint and the
+             * next one in the linked list, then we can't continue traversing
+             * the linked list.
+             */
+            endpoint = mEndpoints.GetHead();
+            continue;
```

Comment:
Should we also reset the `earliestPendingTimerExpiry` back to `now.GetDistantFuture()`?

We are going to go through the list again, and the value in `earliestPendingTimerExpiry ` that we are tracking at this point, may be from an end-point that is now removed from the list. So restting it makes sense to me.

Reply:
Sure, we can do that as well. I restructured this using a `goto` to avoid duplicating initialization logic --- I know using `goto` is typically discouraged but in this case I really do think that it helps simplify things.